x/ibc/06-solomachine/misbehaviour.go

```diff
@@ -0,0 +1,69 @@
+package solomachine
+
+import (
+	"bytes"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	clientexported "github.com/cosmos/cosmos-sdk/x/ibc/02-client/exported"
+	clienttypes "github.com/cosmos/cosmos-sdk/x/ibc/02-client/types"
+)
+
+// CheckMisbehaviourAndUpdateState determines whether or not the currently registered
+// public key signed over two different messages with the same sequence. If this is true
+// the client state is updated to a frozen status.
+func CheckMisbehaviourAndUpdateState(
+	clientState clientexported.ClientState,
+	_ clientexported.ConsensusState,
+	misbehaviour clientexported.Misbehaviour,
+) (clientexported.ClientState, error) {
+
+	// cast the interface to specific types before checking for misbehaviour
+	smClientState, ok := clientState.(ClientState)
+	if !ok {
+		return nil, sdkerrors.Wrapf(clienttypes.ErrInvalidClientType, "client state type %T is not solo machine", clientState)
+	}
+
+	if smClientState.IsFrozen() {
+		return nil, sdkerrors.Wrapf(clienttypes.ErrClientFrozen, "client is already frozen")
+	}
+
+	evidence, ok := misbehaviour.(Evidence)
+	if !ok {
+		return nil, sdkerrors.Wrapf(clienttypes.ErrInvalidClientType, "evidence type %T is not solo machine", misbehaviour)
+	}
+
+	if err := checkMisbehaviour(smClientState, evidence); err != nil {
+		return nil, err
+	}
+
+	smClientState.Frozen = true
+	return smClientState, nil
+}
+
+// checkMisbehaviour checks if the currently registered public key has signed
+// over two different messages at the same sequence.
+func checkMisbehaviour(clientState ClientState, evidence Evidence) error {
+	pubKey := clientState.ConsensusState.PubKey
```

Comment:
At the moment, the spec does not actually do this - https://github.com/cosmos/ics/blob/master/spec/ics-006-solo-machine-client/README.md#misbehaviour-predicate - although it might be a good idea - I don't really understand the solo client "threat model" yet - https://github.com/cosmos/ics/issues/427.

However, even if we do support this, we only need to store each unique public key once along with a sequence range, we don't need to store the consensus state every sequence.

Reply:
The code has been adjusted to use the pubkey in the passed in consensus state which is retrieved using the [`GetClientConsensusStateLTE`](https://github.com/cosmos/cosmos-sdk/blob/257354dbff70197420a386febea5be4df5d71660/x/ibc/02-client/keeper/client.go#L134) - so this is not an issue anymore